File: backend/app/schemas/subscription.py
"""
Subscription Schemas

Pydantic models for payment/subscription requests and responses.
"""

from pydantic import BaseModel
from typing import Optional
from datetime import datetime
from enum import Enum


class SubscriptionPlan(str, Enum):
    """Available subscription plans"""
    MONTHLY = "monthly"
    YEARLY = "yearly"


class SubscriptionStatus(str, Enum):
    """Subscription status values"""
    ACTIVE = "active"
    CANCELLED = "cancelled"
    PAST_DUE = "past_due"
    TRIALING = "trialing"
    INCOMPLETE = "incomplete"
    NONE = "none"


# =============================================================================
# Request Schemas
# =============================================================================

class CreateCheckoutRequest(BaseModel):
    """Request to create a Stripe checkout session"""
    plan: SubscriptionPlan
    promo_code: Optional[str] = None


class CancelSubscriptionRequest(BaseModel):
    """Request to cancel subscription"""
    cancel_at_period_end: bool = True  # Cancel at end of billing period


class ApplyPromoCodeRequest(BaseModel):
    """Request to validate/apply a promo code"""
    promo_code: str


# =============================================================================
# Response Schemas
# =============================================================================

class CheckoutSessionResponse(BaseModel):
    """Response with Stripe checkout URL"""
    checkout_url: str
    session_id: str


class SubscriptionResponse(BaseModel):
    """Current subscription status"""
    status: SubscriptionStatus
    plan: Optional[SubscriptionPlan] = None
    current_period_end: Optional[datetime] = None
    cancel_at_period_end: bool = False
    stripe_customer_id: Optional[str] = None


class PromoCodeResponse(BaseModel):
    """Promo code validation response"""
    valid: bool
    discount_percent: Optional[int] = None
    discount_amount: Optional[float] = None
    message: str


class CustomerPortalResponse(BaseModel):
    """Stripe Customer Portal URL"""
    portal_url: str


class WebhookResponse(BaseModel):
    """Webhook processing response"""
    received: bool
    message: str = "Webhook processed"

File: frontend/src/services/api/subscription.ts
/**
 * Subscription API Module
 *
 * Handles all Stripe subscription-related API calls.
 */

import apiClient from './client';

// Subscription types
export type SubscriptionPlan = 'monthly' | 'yearly';

export type SubscriptionStatus =
  | 'active'
  | 'cancelled'
  | 'past_due'
  | 'trialing'
  | 'incomplete'
  | 'none';

export interface SubscriptionInfo {
  status: SubscriptionStatus;
  plan?: SubscriptionPlan;
  current_period_end?: string;
  cancel_at_period_end: boolean;
  stripe_customer_id?: string;
}

export interface CheckoutResponse {
  checkout_url: string;
  session_id: string;
}

export interface PromoCodeResponse {
  valid: boolean;
  discount_percent?: number;
  discount_amount?: number;
  message: string;
}

interface CustomerPortalResponse {
  portal_url: string;
}

export const subscriptionApi = {
  /**
   * Get current subscription status
   */
  getStatus: async (): Promise<SubscriptionInfo> => {
    const response = await apiClient.get<SubscriptionInfo>('/subscriptions/status');
    return response.data;
  },

  /**
   * Create checkout session for subscription
   */
  createCheckout: async (
    plan: SubscriptionPlan,
    promoCode?: string
  ): Promise<CheckoutResponse> => {
    const response = await apiClient.post<CheckoutResponse>('/subscriptions/checkout', {
      plan,
      promo_code: promoCode,
    });
    return response.data;
  },

  /**
   * Cancel current subscription
   */
  cancel: async (cancelAtPeriodEnd: boolean = true): Promise<SubscriptionInfo> => {
    const response = await apiClient.post<SubscriptionInfo>('/subscriptions/cancel', {
      cancel_at_period_end: cancelAtPeriodEnd,
    });
    return response.data;
  },

  /**
   * Reactivate cancelled subscription
   */
  reactivate: async (): Promise<SubscriptionInfo> => {
    const response = await apiClient.post<SubscriptionInfo>('/subscriptions/reactivate');
    return response.data;
  },

  /**
   * Get customer portal URL for self-service
   */
  getPortalUrl: async (): Promise<string> => {
    const response = await apiClient.post<CustomerPortalResponse>('/subscriptions/portal');
    return response.data.portal_url;
  },

  /**
   * Validate promo code
   */
  validatePromoCode: async (code: string): Promise<PromoCodeResponse> => {
    const response = await apiClient.post<PromoCodeResponse>('/subscriptions/validate-promo', {
      promo_code: code,
    });
    return response.data;
  },
};
